### PR TITLE
Fixes #9250 - Add article showing how to invoke a PSCmdlet

### LINKED ISSUE #9250: May want to call out what to do to invoke a command that derives `PSCmdlet`
### Prerequisites

- [X] **Existing Issue:** Search the existing issues for this repository. If there is an issue that fits your needs do not file a new one. Subscribe, react, or comment on that issue instead.
- [X] **Descriptive Title:** Write the title for this issue as a short synopsis. If possible, provide context. For example, "Typo in `Get-Foo` cmdlet" instead of "Typo."
- [X] **Verify Version:** If there is a mismatch between documentation and the behavior on your system, ensure that the version you are using is the same as the documentation. Check this box if they match or the issue you are reporting is not version specific.

### Links

https://learn.microsoft.com/en-us/powershell/scripting/developer/cmdlet/how-to-invoke-a-cmdlet-from-within-a-cmdlet?view=powershell-7.2

### Summary

As https://github.com/PowerShell/PowerShell/issues/18159 described:

> There's an official article describing how to invoke a Cmdlet from another cmdlet: https://learn.microsoft.com/en-us/powershell/scripting/developer/cmdlet/how-to-invoke-a-cmdlet-from-within-a-cmdlet?view=powershell-7.2
However, it does not describe any way to do the same for PSCmdlet, which I believe is a much more common request

The NOTE message in that page says:
> You can invoke only those cmdlets that derive directly from the [System.Management.Automation.Cmdlet](https://learn.microsoft.com/en-us/dotnet/api/System.Management.Automation.Cmdlet) class. You cannot invoke a cmdlet that derives from the [System.Management.Automation.PSCmdlet](https://learn.microsoft.com/en-us/dotnet/api/System.Management.Automation.PSCmdlet) class.

Maybe we should point to a reference about how to invoke `PSCmdlet` from another `PSCmdlet` here.

### Details

_No response_

### Suggested Fix

See https://github.com/PowerShell/PowerShell/issues/18159#issuecomment-1258583500 and https://github.com/PowerShell/PowerShell/issues/18159#issuecomment-1260188171 (quoted here: "_Broadly speaking, the only practical way to do cmdlets calling cmdlets is to use the Pipeline API._")

### PR DESCRIPTION
**Pull request by sdwheeler**
# PR Summary

Fixes [AB#13011](https://dev.azure.com/msft-skilling/cebd7ef5-4282-448b-9701-88c8637581b7/_workitems/edit/13011) - Fixes #9250 - Add article showing how to invoke a PSCmdlet

## PR Checklist

- [x] **Descriptive Title:** This PR's title is a synopsis of the changes it proposes.
- [x] **Summary:** This PR's summary describes the scope and intent of the change.
- [x] **Contributor's Guide:** I have read the [contributors guide][contrib].
- [x] **Style:** This PR adheres to the [style guide][style].

[contrib]: https://learn.microsoft.com/powershell/scripting/community/contributing/overview
[style]: https://learn.microsoft.com/powershell/scripting/community/contributing/powershell-style-guide

### CODE REVIEW

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: reference/docs-conceptual/developer/cmdlet/how-to-invoke-a-pscmdlet-from-within-a-pscmdlet.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ms.date: 10/06/2022
+ms.topic: reference
+title: How to invoke a PSCmdlet from within a PSCmdlet
+---
+# How to invoke a PSCmdlet from within a PSCmdlet
+
+This example shows how to invoke a PSCmdlet from within another PSCmdlet. In this example, the new
+PSCmdlet `Get-ClipboardReverse` calls `Get-Clipboard` to get the contents of the clipboard. The
+`Get-ClipboardReverse` reverses the order of the characters and returns the reversed string.
+
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/developer/cmdlet/how-to-invoke-a-pscmdlet-from-within-a-pscmdlet.md:18]: See the note I added.

**Review by SteveL-MSFT — APPROVED** [review]
LGTM

**Comment by MatejKafka** [comment]
@sdwheeler In the article, the PowerShell class is used. From my tests, the invocation is quite slow that way. In the meantime, I found another way to invoke a PSCmdlet:
```powershell
function fn {
    [CmdletBinding()]param()
    $ss = $PSCmdlet.SessionState
    $ss.InvokeCommand.InvokeScript($ss, [ScriptBlock]::Create('Get-Clipboard -Raw'))
}
```
(it works the same from a C# binary cmdlet)

This way of invoking `Get-Clipboard` is about 5 times faster for me (2ms vs 10ms for the `PowerShell.Create` approach). Am I missing any downside of this approach?

**Comment by SeeminglyScience** [comment]
> @sdwheeler In the article, the PowerShell class is used. From my tests, the invocation is quite slow that way. In the meantime, I found another way to invoke a PSCmdlet:
> 
> ```powershell
> function fn {
>     [CmdletBinding()]param()
>     $ss = $PSCmdlet.SessionState
>     $ss.InvokeCommand.InvokeScript($ss, [ScriptBlock]::Create('Get-Clipboard -Raw'))
> }
> ```
> 
> (it works the same from a C# binary cmdlet)
> 
> This way of invoking `Get-Clipboard` is about 5 times faster for me (2ms vs 10ms for the `PowerShell.Create` approach). Am I missing any downside of this approach?

The doc should be using `PowerShell.Create(RunspaceMode.CurrentRunspace)`. That's likely the problem you're hitting in your benchmark, because otherwise it spins up a new runspace for every invocation.

**Comment by MatejKafka** [comment]
@SeeminglyScience 

With `RunspaceMode.CurrentRunspace`, it is significantly faster, but still slower than `InvokeScript`, which I find a bit surprising, given that `InvokeScript` has to first parse the script-block.

My naive benchmark code:
```csharp
using System.Diagnostics;
using System.Management.Automation;

namespace RandomTests;

[Cmdlet(VerbsDiagnostic.Test, "PSCmdlet")]
public class Test : PSCmdlet {
    public string PowerShellCreate() {
        var ps = PowerShell.Create(RunspaceMode.CurrentRunspace);
        ps.AddCommand("Get-Clipboard").AddParameter("-Raw");
        return ps.Invoke<string>()[0];
    }

    public string InvokeScript() {
        var sb = ScriptBlock.Create("Get-Clipboard -Raw");
        return (string)SessionState.InvokeCommand.InvokeScript(SessionState, sb)[0].BaseObject;
    }

    protected override void BeginProcessing() {
        base.BeginProcessing();
        var sw = new Stopwatch();
        sw.Start();
        for (var i = 0; i < 10000; i++) {
            PowerShellCreate();
        }
        sw.Stop();
        WriteObject($"PowerShellCreate: {sw.ElapsedMilliseconds}");
        sw.Restart();
        for (var i = 0; i < 10000; i++) {
            InvokeScript();
        }
        sw.Stop();
        WriteObject($"InvokeScript: {sw.ElapsedMilliseconds}");
    }
}
```

Results:
```
> for ($i = 0; $i -lt 10; $i++) {Test-PSCmdlet}
PowerShellCreate: 618
InvokeScript: 369
PowerShellCreate: 539
InvokeScript: 357
PowerShellCreate: 588
InvokeScript: 407
PowerShellCreate: 626
InvokeScript: 404
PowerShellCreate: 632
InvokeScript: 402
PowerShellCreate: 616
InvokeScript: 409
PowerShellCreate: 614
InvokeScript: 405
PowerShellCreate: 615
InvokeScript: 407
PowerShellCreate: 620
InvokeScript: 418
PowerShellCreate: 619
InvokeScript: 424
```

**Comment by SeeminglyScience** [comment]
I'd use [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet) for the tests, but either way the `PowerShell` API will more consistent and typically more natural to use from C#.

**Pull request merged**